### PR TITLE
Add enumerables module

### PR DESCRIPTION
**Pull request by Aduda-Boaz**
This is a simple Ruby enumerable method implementation exercise, done collaboratively by @omar-labana and @Aduda-Boaz. 
We approach the exercise by implementing our own methods. The project exercise is built with pure Ruby, following the git workflows.
We are proud to present our work and we welcome any feedback that will help us improve on our work and even master the language. This project is purely for our learning purposes and hopes to implement it in the future.

### CODE REVIEW

**Review by GzimAsani — CHANGES_REQUESTED** [review]
# STATUS: Changes Required :red_circle: :pray: 

## Hello @Aduda-Boaz and @omar-labana you have done great job so far with your project however we have few things to improve in order to complete this project, kindly find them below :arrow_down: 

- [x] my_each
- returns an Enumerator if no block is given
- when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- when a block is given when self is a range returns the range itself after calling the given block once for each element in self  expected: 5..50 got: 46
- when a block is given when self is a hash behaves like #each calls the given block once for each element in self expected: "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"   got: "a1b2c3d4e5"
- [x] my_each_with_index
- returns an enumerator if no block is given LocalJumpError:  no block given (yield)
- when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- when a block is given when self is a range returns the range itself after calling the given block once for each element in self expected: 5..50 got: 46
- when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self
- [x] my_select
- returns an enumerator if no block is given
- [x] my_all?
- when no block or argument is given returns true when none of the collection members are false or nil
      Failure/Error: my_each { |element| return false unless yield element }
- when no block or argument is given returns false when one of the collection members are false or nil
      Failure/Error: my_each { |element| return false unless yield element }
- when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
      Failure/Error:
        def my_all?
          my_each { |element| return false unless yield element }
          true
        end
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
- when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
      Failure/Error:
        def my_all?
          my_each { |element| return false unless yield element }
          true
        end
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
- when a class is passed as an argument returns false if any of the collection is not a member of such class
      Failure/Error:
        def my_all?
          my_each { |element| return false unless yield element }
          true
        end
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
      Failure/Error:
        def my_all?
          my_each { |element| return false unless yield element }
          true
        end
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
- when a Regex is passed as an argument returns false if any of the collection does not match the Regex
      Failure/Error:
        def my_all?
          my_each { |element| return false unless yield element }
          true
        end
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
- when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
      Failure/Error:
        def my_all?
          my_each { |element| return false unless yield element }
          true
        end
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
- when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
      Failure/Error:
        def my_all?
          my_each { |element| return false unless yield element }
          true
        end
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
- [x] my_any?
- when no block or argument is given returns true if at least one of the collection is not false or nil
-  when no block or argument is given returns false if at least one of the collection is not true
- when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
      Failure/Error:
        def my_any?
          my_each { |element| return true if yield element }
          false
        end
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
- when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
- when a Regex is passed as an argument returns true if any of the collection matches the Regex
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
- [x] my_none?
- when no block or argument is given returns true only if none of the collection members is true
- when no block or argument is given returns false only if one of the collection members is true
      Failure/Error: my_each { |element| return false if yield element }
      
      LocalJumpError:
        no block given (yield)
- when a class is passed as an argument returns true if none of the collection is a member of such class
- when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
- when a class is passed as an argument returns false if any of the collection is a member of such class
- when a Regex is passed as an argument returns true if none of the collection matches the Regex
- when a Regex is passed as an argument returns false if any of the collection matches the Regex
- when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
- [x] my_count
- returns the number of items in enum through enumeration
      Failure/Error: my_each { |element| count += 1 if yield element }
      
      LocalJumpError:
        no block given (yield)
- returns the number of items in enum through enumeration::range
- counts the number of items in enum that are equal to item if an argument is given
- [x] my_map
- returns an Enumerator if no block is given
      Failure/Error: my_each { |element| pivot_array << yield(element) }
      
      LocalJumpError:
        no block given (yield)
- [ ] my_inject
- when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
      Failure/Error:
        self[start_index...length].my_each do |element|
          total = yield(total, element)
        end
      
      NameError:
        undefined local variable or method `length' for 5..50:Range
-  when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
- when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
- when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
- when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
      Failure/Error:
        def my_inject(inject_parameter = nil)
          if inject_parameter
            total = inject_parameter
            start_index = 0
          else
            total = first
            start_index = 1
          end
          self[start_index...length].my_each do |element|
            total = yield(total, element)
      
      ArgumentError:
        wrong number of arguments (given 2, expected 0..1)
- when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
- accepts an array as an argument and multiplies all the elements of the array together using #my_inject

_Please implement all the above mentioned requirements and ask for another code review whenever you are ready_

### **If you have any question related to the requirements please add a comment here and ping me on slack @GzimAsani**

# HAPPY CODING :computer: :desktop_computer:

**Comment by omar-labana** [comment]
Please paste the test cases if any changes are required, Thanks.

**Review by BrianSammit — CHANGES_REQUESTED** [review]
CHANGES REQUIRED :red_circle: :stop_sign: 
------------
Hello, @Aduda-Boaz @omar-labana you are doing an excellent job so far. :muscle: 
However, there are a few changes to make to complete this project. :hammer_and_wrench: :see them below :eyes: :point_down: 
---------
- [x] #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self. It was expecting `expected: "[:a, 1][:b, 2][:c, 3][:d, 4][:e, 5]"` instead `got: "abcde"`
- [x] #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self. It was expecting `expected: "Num: [:a, 1], idx: 0\nNum: [:b, 2], idx: 1\nNum: [:c, 3], idx: 2\nNum: [:d, 4], idx: 3\nNum: [:e, 5], idx: 4\n"` instead `got: "Num: a, idx: 1\nNum: b, idx: 2\nNum: c, idx: 3\nNum: d, idx: 4\nNum: e, idx: 5\n"`
- [x] #my_any? returns true if the block ever returns a value other than false or nil. `expected: true` instead `got: false`
- [x] #my_any? returns false if the block ever returns a value false or nil. `expected: false` instead `got: true`
- [x] #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil. `expected true` instead `got false`
- [x] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern. `expected true` instead `got false`
- [ ] #my_none? returns true if the block never returns true for all elements. I got an error `NameError: undefined local variable or method `nill'` You just misspell nil. 
- [ ] #my_inject searches for the longest word in an array of strings. I got an error `NoMethodError: undefined method `length' for nil:NilClass`
- [ ] #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block. I got an error `TypeError: nil can't be coerced into Integer`
- [ ] #my_inject when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range. I got an error `TypeError: nil can't be coerced into Integer`
- [ ] #my_inject when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method. I got an error `ArgumentError: tried to create Proc object without a block `
- [ ] You are missing the `#multiply_els` method. please add it. 
--------------
Make these required changes and send it again for another review. :clap: :+1: 
HAPPY CODING!! :smile:

**Review by mricanho — CHANGES_REQUESTED** [review]
# 🔴  Changes Required 🔴  

Hello @Aduda-Boaz @omar-labana, you did a good job so far🔥, just a few changes are required:
## my_each
- [ ]   Returns an Enumerator if no block is given 

```
ArgumentError:
       wrong number of arguments (given 1, expected 0)
```

## my_each_with_index

- [ ] Returns an enumerator if no block is given
```
ArgumentError:
       wrong number of arguments (given 1, expected 0)
```

## my_any?

- [ ] When a class is passed as an argument returns true if at least one of the 
collection is a member of such class::Numeric

```
expected true
            got false
```

- [ ] When a Regex is passed as an argument returns true if any of the collection matches the Regex

```
expected true
            got false
```
## my_none?

- [ ] When a Regex is passed as an argument returns false if any of the collection matches the Regex

```
expected false
            got true
```

## my_count

- [ ] Counts the number of items in enum that are equal to item if an argument 
is given

```
LocalJumpError:
       no block given (yield)
```

## my_inject

- [ ] Raises a "LocalJumpError" when no block or argument is given

`expected LocalJumpError, got #<ArgumentError`

- [ ] Searches for the longest word in an array of strings

```
NoMethodError:
       undefined method `length' for nil:NilClass
```

- [ ]  When a block is given without an initial value combines all elements of 
enum by applying a binary operation, specified by a block

```
TypeError:
       nil can't be coerced into Integer
```

- [ ]  When a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range


```
TypeError:
        nil can't be coerced into Integer
```

- [ ]   When a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range

```
TypeError:
       nil can't be coerced into Integer
```

- [ ] When a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block

```
TypeError:
        nil can't be coerced into Integer
```

- [ ] When a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method

```
ArgumentError:
        tried to create Proc object without a block
```

- [ ] When a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range

```
ArgumentError:
        tried to create Proc object without a block
```

- [ ] When a symbol is specified with an initial value combines each element 
of the collection by applying the symbol as a named method

```
ArgumentError:
        wrong number of arguments (given 2, expected 0..1)
```

- [ ] When a symbol is specified with an initial value combines each element 
of the collection by applying the symbol as a named method::range

```
ArgumentError:
        wrong number of arguments (given 2, expected 0..1)
```

## multiply_els

- [ ] Accepts an array as an argument and multiplies all the elements of the array together using #my_inject

```
TypeError:
        nil can't be coerced into Integer
```

a lot of work

I know it's a lot, but you got this.

You can contact me on slack if you have any questions about how to implement the arrangement for this methods, just search for me as @ Miguel Ricaño.

### 💻 👩🏽‍💻  Happy Coding!👨🏽‍💻  🖥️

**Comment by mimipeshy** [comment]
## STATUS: CHANGES REQUESTED 🔴 ♻️ 

Hello team @Aduda-Boaz and @omar-labana 👋🏾 👋🏾 , you guys have done an amazing job with the feedback you have been given. Keep it up 👏🏾 👏🏾 , However, let's make one more change before we can approve this project  ⬇️ 

- [x] Check out the comment I will leave below


That's all. 😄 , kindy submit another review when you are done

Regards
Peris Ndanu
_If you have any questions, kindly leave a comment here and tag my name @mimipeshy and I will be glad to help_

File: lib/enumerable_module.rb
Comment (mimipeshy) [inline lib/enumerable_module.rb]: - [x] enumerables #my_count counts the number of elements yielding a true value if a block is given

**Review by taiwo2 — APPROVED** [review]
# Status: Approved :tada: 

Hello @Aduda-Boaz and @omar-labana  Weldone for the Job done so far, you Make your Efforts commendable, never relent and prepare for your next project.

**Pull request merged**
